Add support for transferring badge config via Bluetooth Low Energy

hlxid opened the pull request:
Adds a new crate feature `ble` that allows for transfering badge configurations over Bluetooth Low Energy instead of USB. Expands the CLI to allow for specification of the transport protocol, either USB or BLE.

The BLE library uses async rust requiring to add tokio as a async runtime to the CLI.

<!-- Generated by sourcery-ai[bot]: start summary -->

## Summary by Sourcery

This pull request introduces support for transferring badge configurations via Bluetooth Low Energy (BLE) by adding a new crate feature `ble`. The CLI has been expanded to allow users to specify the transport protocol (USB or BLE). Additionally, the `tokio` async runtime has been integrated to support the asynchronous operations required by the BLE library. Documentation has been updated to reflect these changes.

- **New Features**:
    - Added support for transferring badge configurations via Bluetooth Low Energy (BLE) by introducing a new crate feature `ble`.
    - Expanded the CLI to allow specification of the transport protocol, either USB or BLE.
- **Enhancements**:
    - Integrated the `tokio` async runtime to support asynchronous operations required by the BLE library.
- **Documentation**:
    - Updated the README to include instructions for using the new BLE transport mode, including necessary permissions for macOS.

<!-- Generated by sourcery-ai[bot]: end summary -->

mgjm commented on src/main.rs:
```diff
@@ -177,7 +190,10 @@ fn main() -> Result<()> {
-    Device::single()?.write(payload)?;
+    match args.transport {
+        TransportProtocol::Usb => UsbDevice::single()?.write(payload),
+        TransportProtocol::Ble => BleDevice::single().await?.write(payload).await,
```
Let's keep the `main` function blocking and start an async runtime only when needed:
```suggestion
        TransportProtocol::Ble => tokio::runtime::Builder::new_current_thread()
            .enable_all()
            .build()?
            .block_on(async {
                BleDevice::single().await?.write(payload).await
            }),
```

Note: The `macros` feature of `tokio` is then no longer needed.

mgjm reviewed (changes requested):
Thanks for the PR ❤️. BLE support was on my wishlist.

I have written a few improvement suggestions. In general please have look at all `?` operators. Every time you return an error from a library like `btleplug` try to add a custom context to the error. This helps to narrow down where an error originated from (kind of like a stack trace). I have given one example suggestion, but this can be applied to all `?` operators.

Note: All suggestions are just written in GitHub, they still need `cargo fmt` and maybe some `use` statements are missing.

I can't test the BLE functionality myself. Did you test this feature on real hardware with an actual badge? And on which OS?

mgjm commented:
The CI pipeline failed on linux. The `dbus-1` library seems to be missing. Try to add the `libdbus-1-dev` package in [`.github/workflows/ci.yaml`](https://github.com/fossasia/badgemagic-rs/blob/main/.github/workflows/ci.yaml#L54) at line `54` and `69`.

hlxid commented:
Thank you very much for your extensive review! All addressed points should be fixed with my latest commit.

I have tested this PR with a real badge on two different Arch Linux machines. I can also test it on Windows and macOS if you want.

mgjm commented on src/ble.rs:
```diff
@@ -0,0 +1,179 @@
+/// `0000fee0-0000-1000-8000-00805f9b34fb`
+const BADGE_SERVICE_UUID: Uuid = btleplug::api::bleuuid::uuid_from_u16(0xfee0);
+/// `0000fee1-0000-1000-8000-00805f9b34fb`
+const BADGE_CHAR_UUID: Uuid = btleplug::api::bleuuid::uuid_from_u16(0xfee1);
```
I guess that's what I wrote in my suggestion, but it would be a bit nicer to read with an `use btleplug::api::bleuuid` statement 🙈:
```suggestion
const BADGE_CHAR_UUID: Uuid = bleuuid::uuid_from_u16(0xfee1);
```

mgjm commented on src/ble.rs:
```diff
@@ -0,0 +1,179 @@
+        // and also the correct name.
+        // The service uuid is also by devices that are not LED badges, so
+        // the name check is also necessary.
+        let props = peripheral.properties().await;
+        if props.is_err() {
+            return None;
+        }
+
+        if let Some(props) = props.unwrap() {
```
Let's eliminate the last two unwraps. Unwrap 1/2:
```suggestion
        let props = peripheral.properties().await.ok()??;
```

The first `.ok()?` converts the `Result` into an `Option` and returns `None` if there was an `Err`. The second `?` returns `None` if the inner `Option` was `None`.

mgjm commented on src/ble.rs:
```diff
@@ -0,0 +1,179 @@
+            .await
+            .context("discovering services")?;
+        let characteristics = self.peripheral.characteristics();
+        let badge_char = characteristics.iter().find(|c| c.uuid == BADGE_CHAR_UUID);
+
+        if badge_char.is_none() {
+            return Err(anyhow::anyhow!("Badge characteristic not found"));
+        }
+        let badge_char = badge_char.unwrap();
```
Unwrap 2/2. The `anyhow::Context` trait also work for `Option`s:
```suggestion
        let badge_char = characteristics.iter().find(|c| c.uuid == BADGE_CHAR_UUID).context("badge characteristic not found")?;
```

mgjm commented on src/ble.rs:
```diff
@@ -0,0 +1,179 @@
+            .context("bluetooth device connect")?;
+
+        let result = self.write_connected(payload).await;
+        if result.is_ok() {
+            self.peripheral
+                .disconnect()
+                .await
+                .context("bluetooth device disconnect")?;
+        }
```
We want to always disconnect the peripheral, but ignore any errors during disconnect if we are already in an error state:
```suggestion
        let disconnect_result = self.peripheral
            .disconnect()
            .await;
        if result.is_ok() {
            disconnect_result.context("bluetooth device disconnect")?;
        } else {
            // ignore result of disconnect (an error already occurred)
        }
```

mgjm reviewed (changes requested):
Awesome. I've just added some last suggestions.

And it would be great if you could test it on Windows and MacOS too.

Note: I haven't tested the code on Windows or MacOS at all. There could be unrelated bugs. If you have the time, you could also test the USB transport on those systems.

mgjm commented on src/ble.rs:
```diff
@@ -0,0 +1,179 @@
+        let bytes = payload.into_padded_bytes();
+        let data = bytes.as_ref();
+
+        anyhow::ensure!(
+            data.len() % BLE_CHAR_CHUNK_SIZE == 0,
+            format!(
+                "Payload size must be a multiple of {} bytes",
+                BLE_CHAR_CHUNK_SIZE
+            )
+        );
```
The `anyhow::ensure` macro already supports `format`-like formatting:
```suggestion
        anyhow::ensure!(
            data.len() % BLE_CHAR_CHUNK_SIZE == 0,
            "Payload size must be a multiple of {} bytes",
            BLE_CHAR_CHUNK_SIZE,
        );
```

hlxid commented:
Thanks again. I've also removed the service UUID check as it is done through the `ScanFilter` now and would be redundant.

I've tested this now on macOS where writing to the badge works but afterwards the CLI just decides to hang indefinitely and not exit as expected. USB works flawlessly :)
I will be able to look into the hang issue after writing on macOS and test on Windows sometimes in the beginning of next week.

mgjm commented:
LGTM

I've moved the MacOS and Windows testing to a separate issue: #20
